### PR TITLE
Remove the old FOLLOW checking (aka `check_matcher_old`).

### PR DESCRIPTION
**Pull request by LeoTestard**
It was supposed to be removed at the next release cycle but is still in the tree since like 6 months.
Potential breaking change, since some cases (such as #25658) will change from a warning to an error. But the warning stating that it will be a hard error in the next release has been there for 6 months now.
I think it's safe to break this code. ^_^

### CODE REVIEW

**Comment by pnkfelix** [comment]
triage: nominated

(Nominating since I just want to make sure lang team is aware of the long awaited breaking change here, and also because I'm not 100% sure about the protocol here.)

**Comment by nikomatsakis** [comment]
Conclusion from lang-team meeting: let's follow a foreshortened protocol here, but at least post a  warning to internals.

**Comment by pnkfelix** [comment]
posted warning to internals: https://internals.rust-lang.org/t/notice-foreshortened-warning-hard-error-plan-for-33982/3543

**Comment by pnkfelix** [comment]
cc #30450

**Comment by durka** [comment]
@LeoTestard does this build locally for you? I'm trying to hack on it and getting errors in libcore that suggests macros are silently failing to expand.

File: src/libsyntax/ext/tt/macro_rules.rs [hunk 1]
```diff
@@ -354,203 +354,12 @@ fn check_rhs(cx: &mut ExtCtxt, rhs: &TokenTree) -> bool {
...
-    last
-}
-
-fn check_matcher_new(cx: &mut ExtCtxt, matcher: &[TokenTree], on_fail: &mut OnFail) {
     let first_sets = FirstSets::new(matcher);
     let empty_suffix = TokenSet::empty();
-    check_matcher_core(cx, &first_sets, matcher, &empty_suffix, on_fail);
+    let err = cx.parse_sess.span_diagnostic.err_count();
+    check_matcher_core(cx, &first_sets, matcher, &empty_suffix);
+    err < cx.parse_sess.span_diagnostic.err_count()
```
Comment (durka) [inline src/libsyntax/ext/tt/macro_rules.rs]: This check is backwards! `<` should be `==`. 😆

Comment (LeoTestard) [inline src/libsyntax/ext/tt/macro_rules.rs]: Ooooh right, I'm so stupid. ><

**Comment by LeoTestard** [comment]
@durka: I must admit that I haven't tested it and I was waiting for Travis to do it. The error you pointed out is probably the cause. The fact that it fails silently is legit because the errors are supposed to be reported directly by `check_matcher_core`, when present. But the return value of `false` will just cause the macro to be compiled in a dummy macro that expands to nothing to avoid duplicate errors. I will fix that soon. :)

**Comment by durka** [comment]
You need to also replace all instances of `WARN` with `ERROR` in src/test/compile-fail/macro-follow.rs.

**Comment by durka** [comment]
Actually the macro in issue-30715.rs is no longer legal with these changes. Maybe we should just delete that test?

**Comment by brson** [comment]
Looks like this broke a number of crates https://internals.rust-lang.org/t/regression-report-stable-2016-05-24-vs-nightly-2016-06-10/3594/6

**Pull request merged**